src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseStream.cs

```diff
@@ -112,11 +106,19 @@ public override Task CopyToAsync(Stream destination, int bufferSize, Cancellatio
 
         private async Task CopyToAsyncCore(Stream destination, byte[] buffer, CancellationToken cancellationToken)
         {
-            _state.PinReceiveBuffer(buffer);
             CancellationTokenRegistration ctr = cancellationToken.Register(s => ((WinHttpResponseStream)s!).CancelPendingResponseStreamReadOperation(), this);
-            _state.AsyncReadInProgress = true;
+            lock (_state.Lock)
+            {
+                // Check that there are no other pending read operations
+                if (_state.AsyncReadInProgress)
+                {
+                    throw new InvalidOperationException(SR.net_http_no_concurrent_io_allowed);
+                }
+                _state.AsyncReadInProgress = true;
+            }
```

Comment:
Could we instead make AsyncReadInProgress a volatile int field and change this (and the other occurrence of it) to be:
```C#
if (Interlocked.ComparerExchange(ref _state.AsyncReadInProgress, 1, 0) != 0)
{
    throw new InvalidOperationException(SR.net_http_no_concurrent_io_allowed);
}
```
?

Reply:
Yep, I even wrote it in the description as another way to fix this.